lib/scholar/neighbors/k_nearest_neighbors.ex

```diff
@@ -0,0 +1,401 @@
+  defstruct [:data, :labels, :default_num_neighbors, :weights, :num_classes, :p, :task, :metric]
+
+  opts = [
+    num_neighbors: [
+      type: :pos_integer,
+      default: 5,
+      doc: "The number of neighbors to use by default for `k_neighbors` queries"
+    ],
+    num_classes: [
+      type: :pos_integer,
+      required: true,
+      doc: "Number of classes in provided labels"
+    ],
+    weights: [
+      type: {:in, [:uniform, :distance]},
+      default: :uniform,
+      doc: """
+      Weight function used in prediction. Possible values:
+
+        * `:uniform` - uniform weights. All points in each neighborhood are weighted equally.
+
+        * `:distance` - weight points by the inverse of their distance. in this case, closer neighbors of
+          a query point will have a greater influence than neighbors which are further away.
+      """
+    ],
+    p: [
+      type: {:custom, Scholar.Options, :positive_number, []},
+      default: 2,
+      doc: """
+      Power parameter for the Minkowski metric. When p = 1, this is equivalent
+      to using manhattan_distance ($L_1$), and euclidean_distance ($L_2$) for p = 2.
+      """
+    ],
+    metric: [
+      type: {:in, [:minkowski, :cosine]},
+      default: :minkowski,
+      doc: ~S"""
+      Name of the metric. Possible values:
+      * `:minkowski` - Minkowski metric. By changing value of `:p` parameter we can set Euclidean, Chebyshev, Manhattan or any arbitrary $L_p$ metric.
+
+      * `:cosine` - Cosine metric.
+      """
+    ],
+    task: [
+      type: {:in, [:classification, :regression]},
+      default: :classification,
+      doc: """
+      Task that will be performed using K Nearest Neighbors. Possible values:
+
+      * `:classification` - Classifier implementing the k-nearest neighbors vote.
+
+      * `:regression` - Regression based on K-Nearest Neighbors.
+        The target is predicted by local interpolation of the targets associated of the nearest neighbors in the training set.
+      """
+    ]
+  ]
+
+  @opts_schema NimbleOptions.new!(opts)
+
+  @doc """
+  Fit the K-nearest neighbors classifier from the training data set.
+
+  For classification, provided labels needs to be consecutive natural numbers. If your labels does
+  not meet this condition please use `Scholar.Preprocessing.ordinal_encode`
+
+  Currently 2D labels only supported for regression task.
+
+  ## Options
+
+  #{NimbleOptions.docs(@opts_schema)}
+
+  ## Return Values
+
+    The function returns a struct with the following parameters:
+
+    * `:data` - Training data.
+
+    * `:labels` - Labels of each point.
+
+    * `:default_num_neighbors` - The number of neighbors to use by default for `k_neighbors` queries.
+
+    * `:weights` - Weight function used in prediction.
+
+    * `:num_classes` - Number of classes in provided labels.
+
+  ## Examples
+
+      iex>  Scholar.Neighbors.KNearestNeighbors.fit(Nx.tensor([[1, 2], [2, 4], [1, 3], [2, 5]]), Nx.tensor([1, 0, 1, 1]),
```

Comment:
```suggestion
      iex> foo = Nx.tensor([[1, 2], [2, 4], [1, 3], [2, 5]])
      iex>  Scholar.Neighbors.KNearestNeighbors.fit(foo, Nx.tensor([1, 0, 1, 1]),
```

But please give it a better name to the tensor. :D